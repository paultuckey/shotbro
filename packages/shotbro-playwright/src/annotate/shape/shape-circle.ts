import type {CircleShape} from "./shape-types";
import type {ShotBroBox} from "../annotate-types";
import {ulid} from "../../util/ulid";

const defaultProps: CircleShape = {
    thickness: 4,
    color: 'deeppink',
    translate: 0,
    translateX: 0,
    translateY: 0,
}

export async function renderCircle(elPos: ShotBroBox, rawShape: CircleShape): Promise<string> {
    const shape: CircleShape = Object.assign({}, defaultProps, rawShape);
    const scope = 'circle-' + ulid();
    const id = rawShape.id || scope;
    const thickness = shape.thickness || 0;
    const translateX = shape.translateX ?? shape.translate ?? 0
    const translateY = shape.translateY ?? shape.translate ?? 0
    return `
        <style>
            .circle.${scope} {
                position: fixed;
                top: ${elPos.y + translateY - (thickness / 2)}px;
                left: ${elPos.x + translateX - (thickness / 2)}px;
                width: ${elPos.w}px;
                height: ${elPos.h}px;
                
                border-width: ${thickness}px;
                border-style: solid;
                border-color: ${shape.color};
                border-radius: ${elPos.w / 2}px / ${elPos.h / 2}px;
                box-shadow: ${thickness / 2}px ${thickness / 2}px 2px 0px rgba(0,0,0,0.13);
            }
        </style>
        <div id="${id}" class="circle ${scope}"></div>`;
}
