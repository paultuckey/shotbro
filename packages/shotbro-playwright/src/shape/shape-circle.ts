import type {CircleShape, ShotBroShape} from "./shape-types";
import type {ShotBroBox} from "../annotate/annotate-types";

import {n2dp} from "./shape-utils";

const defaultProps: CircleShape = {
    thickness: 4,
    color: 'deeppink',
    translate: 0,
    translateX: 0,
    translateY: 0,
}

export async function renderCircle(scope: string, elPos: ShotBroBox, rawShape: ShotBroShape): Promise<string> {
    const shape: CircleShape = Object.assign({}, defaultProps, rawShape.circle);
    const id = shape.id || scope;
    const thickness = shape.thickness!;
    const translateX = shape.translateX ?? shape.translate ?? 0;
    const translateY = shape.translateY ?? shape.translate ?? 0;
    const top = elPos.y - (thickness/2);
    const left = elPos.x - (thickness/2);
    return `
        <style>
            .${scope}.circle {
                position: absolute;
                top: ${n2dp(top + translateY)}px;
                left: ${n2dp(left + translateX)}px;
                width: ${elPos.w + thickness}px;
                height: ${elPos.h + thickness}px;
                
                border-width: ${thickness}px;
                border-style: solid;
                border-color: ${shape.color};
                border-radius: ${n2dp(elPos.w / 2)}px / ${n2dp(elPos.h / 2)}px;
            }
        </style>
        <div id="${id}" class="${scope} circle"></div>`;
}
