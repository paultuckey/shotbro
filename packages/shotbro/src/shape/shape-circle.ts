import type {CircleShape, ShotBroBox} from "./shape-types";
import {CssPropertiesHyphen, toStyleAttr} from "./css-util";

const defaultProps: CircleShape = {
  thickness: 4,
  color: 'deeppink',
  translate: 0,
  translateX: 0,
  translateY: 0,
}

export async function renderCircle(elPos: ShotBroBox, rawShape: CircleShape): Promise<string> {
  const shape: CircleShape = Object.assign({}, defaultProps, rawShape);
  const style = toStyleAttr(toCss(elPos, shape));
  return elPos ? `<div style="${style}"></div>` : '';
}

function toCss(elPos: ShotBroBox, shape: CircleShape): CssPropertiesHyphen {
  const thickness = shape.thickness || 0;
  const translateX = shape.translateX ?? shape.translate ?? 0
  const translateY = shape.translateY ?? shape.translate ?? 0
  return {
    position: 'fixed',
    top: `${elPos.y + translateY - (thickness/2)}px`,
    left: `${elPos.x + translateX - (thickness/2)}px`,
    width: `${elPos.w}px`,
    height: `${elPos.h}px`,

    'border-width': `${thickness}px`,
    'border-style': 'solid',
    'border-color': shape.color,
    'border-radius': `${elPos.w / 2}px / ${elPos.h/2}px`,
    'box-shadow': `${thickness / 2}px ${thickness / 2}px 2px 0px rgba(0,0,0,0.13)`
  };
}