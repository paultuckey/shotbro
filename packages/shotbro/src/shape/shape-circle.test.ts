import {renderCircle} from "./shape-circle";
import type {ShotBroBox, CircleShape} from "./shape-types";

describe('Shape Circle', () => {

  test('Simple test', async () => {
    const shape: CircleShape = {
      //thickness: 'Hello World'
    }
    const elPos: ShotBroBox = {x: 1, y: 2, w: 40, h: 50}
    const html = await renderCircle(elPos, shape);
    expect(html).toMatchSnapshot('circle-simple');
  })
});