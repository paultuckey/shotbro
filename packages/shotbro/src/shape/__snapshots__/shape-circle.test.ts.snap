// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape Circle Simple test: circle-simple 1`] = `"<div style="position:fixed;top:0px;left:-1px;width:40px;height:50px;border-width:4px;border-style:solid;border-color:deeppink;border-radius:20px / 25px;box-shadow:2px 2px 2px 0px rgba(0,0,0,0.13);"></div>"`;
